tests/std/tests/P0896R4_views_common/test.cpp

```diff
@@ -0,0 +1,386 @@
+            static_assert(noexcept(views::common(as_const(rng))) == is_noexcept);
+
+            static_assert(same_as<decltype(as_const(rng) | views::common), RC>);
+            static_assert(noexcept(as_const(rng) | views::common) == is_noexcept);
+        }
+        // ... with rvalue argument
+        static_assert(CanViewCommon<remove_reference_t<Rng>> == is_view || enable_borrowed_range<remove_cvref_t<Rng>>);
+        if constexpr (is_view) {
+            constexpr bool is_noexcept = is_nothrow_move_constructible_v<V>;
+            static_assert(same_as<decltype(views::common(move(rng))), V>);
+            static_assert(noexcept(views::common(move(rng))) == is_noexcept);
+
+            static_assert(same_as<decltype(move(rng) | views::common), V>);
+            static_assert(noexcept(move(rng) | views::common) == is_noexcept);
+        } else if constexpr (enable_borrowed_range<Rng>) {
+            using S                    = decltype(ranges::subrange{declval<Rng>()});
+            constexpr bool is_noexcept = noexcept(S{declval<Rng>()});
+
+            static_assert(same_as<decltype(views::common(move(rng))), V>);
+            static_assert(noexcept(views::common(move(rng))) == is_noexcept);
+
+            static_assert(same_as<decltype(move(rng) | views::common), V>);
+            static_assert(noexcept(move(rng) | views::common) == is_noexcept);
+        }
+
+        // ... with const rvalue argument
+        static_assert(CanViewCommon<const remove_reference_t<Rng>> == (is_view && copyable<V>)
+                      || (!is_view && enable_borrowed_range<remove_cvref_t<Rng>>) );
+        if constexpr (is_view && copyable<V>) {
+            constexpr bool is_noexcept = is_nothrow_copy_constructible_v<V>;
+
+            static_assert(same_as<decltype(views::common(move(as_const(rng)))), V>);
+            static_assert(noexcept(views::common(as_const(rng))) == is_noexcept);
+
+            static_assert(same_as<decltype(move(as_const(rng)) | views::common), V>);
+            static_assert(noexcept(as_const(rng) | views::common) == is_noexcept);
+        } else if constexpr (!is_view && enable_borrowed_range<const remove_cvref_t<Rng>>) {
+            using S                    = decltype(ranges::subrange{declval<const remove_cvref_t<Rng>>()});
+            constexpr bool is_noexcept = noexcept(S{declval<const remove_cvref_t<Rng>>()});
+
+            static_assert(same_as<decltype(views::common(move(as_const(rng)))), V>);
+            static_assert(noexcept(views::common(move(as_const(rng)))) == is_noexcept);
+
+            static_assert(same_as<decltype(move(as_const(rng)) | views::common), V>);
+            static_assert(noexcept(move(as_const(rng)) | views::common) == is_noexcept);
+        }
+    }
+
+
+    // Validate deduction guide
+#if !defined(__clang__) && !defined(__EDG__) // TRANSITION, DevCom-1159442
+    (void) 42;
+#endif // TRANSITION, DevCom-1159442
+    if (!is_constant_evaluated()) {

```

Comment:
this makes me sad. I believe I have to be more fine grained as we need to special case for random_access_range

Reply:
It's a bit of a mess: we can't use non-literal variables in `constexpr` functions period. Clang fails to diagnose when the type is dependent: https://godbolt.org/z/xoMGxs. I think I've fixed this by factoring out the icky parts into a separate function.